com.unity.netcode.gameobjects/Runtime/Serialization/BitWriter.cs

```diff
@@ -0,0 +1,206 @@
+using System;
+using System.Runtime.CompilerServices;
+
+namespace Unity.Multiplayer.Netcode
+{
+    /// <summary>
+    /// Helper class for doing bitwise writes for a FastBufferWriter.
+    /// Ensures all bitwise writes end on proper byte alignment so FastBufferWriter doesn't have to be concerned
+    /// with misaligned writes.
+    /// </summary>
+    public ref struct BitWriter
+    {
+        private Ref<FastBufferWriter> m_Writer;
+        private unsafe byte* m_BufferPointer;
+        private readonly int m_Position;
+        private int m_BitPosition;
+#if DEVELOPMENT_BUILD || UNITY_EDITOR
+        private int m_AllowedBitwiseWriteMark;
+#endif
+        private const int k_BitsPerByte = 8;
+
+        /// <summary>
+        /// Whether or not the current BitPosition is evenly divisible by 8. I.e. whether or not the BitPosition is at a byte boundary.
+        /// </summary>
+        public bool BitAligned
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get => (m_BitPosition & 7) == 0;
+        }
+
+        internal unsafe BitWriter(ref FastBufferWriter writer)
+        {
+            m_Writer = new Ref<FastBufferWriter>(ref writer);
+            m_BufferPointer = writer.BufferPointer + writer.PositionInternal;
+            m_Position = writer.PositionInternal;
+            m_BitPosition = 0;
+#if DEVELOPMENT_BUILD || UNITY_EDITOR
+            m_AllowedBitwiseWriteMark = (m_Writer.Value.AllowedWriteMark - m_Writer.Value.Position) * k_BitsPerByte;
+#endif
+        }
+
+        /// <summary>
+        /// Pads the written bit count to byte alignment and commits the write back to the writer
+        /// </summary>
+        public void Dispose()
+        {
+            var bytesWritten = m_BitPosition >> 3;
+            if (!BitAligned)
+            {
+                // Accounting for the partial write
+                ++bytesWritten;
+            }
+
+            m_Writer.Value.CommitBitwiseWrites(bytesWritten);
+        }
+
+        /// <summary>
+        /// Verifies the requested bit count can be written to the buffer.
+        /// This exists as a separate method to allow multiple bit writes to be bounds checked with a single call.
+        /// If it returns false, you may not write, and in editor and development builds, attempting to do so will
+        /// throw an exception. In release builds, attempting to do so will write to random memory addresses and cause
+        /// Bad Things(TM).
```

Comment:
nit but since these API comments go online as a learning resource I'd suggest a more professional "and cause unexpected behavior".